checks.d/zk.py

```diff
@@ -96,6 +166,44 @@ def check(self, instance):
                 self.service_check('zookeeper.mode', status, message=message,
                                    tags=sc_tags)
 
+        if zk_version and LooseVersion(zk_version) > LooseVersion("3.4.0"):
+            try:
+                mntr_out = self._send_command('mntr', *cx_args)
+            except ZKConnectionFailure:
+                self.increment('zookeeper.timeouts')
+                self.report_instance_mode(hostname, 'down', tags)
+                raise
+            except:
+                e = sys.exc_info()[1]
+                print >> sys.stderr, "Error: %s" % e
+                self.increment('zookeeper.datadog_client_exception')
+                self.report_instance_mode(hostname, 'unknown', tags)
+                raise
+            else:
+                metrics, mode = self.parse_mntr(mntr_out)
+                mode_tag = "mode:%s" % mode
+                if mode != 'inactive':
+                    for name in metrics:
+                        self.gauge(name, metrics[name], tags=tags + [mode_tag])
+
+
+    def report_instance_mode(self, hostname, mode, tags):
+        tags = tags + ['mode:%s' % mode]
+        self.set('zookeeper.instances', hostname, tags=tags)
+
+        gauges = {}
+        for key in self.STATUS_TYPES:
+            gauges[key] = 0
+
+        if mode in gauges:
+            gauges[mode] = 1
+        else:
+            gauges['unknown'] = 1
+
+        for k in gauges:
+            gauge_name = 'zookeeper.instances.%s' % k
+            self.gauge(gauge_name, gauges[k])
```

Comment:
@jpittis I wonder if the fact the code functions as you expect is a result of you happening to run the zookeeper instances on the same host as the agent. The `host: <hostname>` tag added by the agent is inferred from the machine the agent runs on. But the way the check is set up right now, the agent need not be monitoring `localhost` but can in fact be monitoring remote hosts with different hostnames. This is why the tag added by the agent can be inaccurate unless you explicitly add additional context. I see your point about the gauges registering correctly, but I wonder whether the check will be more useful in the datadog backend if some explicit tagging were included. I'm going to set up a ZK cluster of my own to play around with your check - hopefully that will get us on the same page :)